### PR TITLE
Fixes #2255, Fixes #2299 API Endpoint for News Export

### LINKED ISSUE #2255: API Endpoint for News Export
## Motivation
We would like to make news content available to external services that wish to import the content programmatically, notably Trellis.

## Proposed Resolution
An API that allows a quickstart site to provide information about its news content.

### Describe the solution you'd like
Investigate options for making news content available - potentially JSON:API or REST services.

### Describe alternatives you've considered
A custom solution for making the content available - potentially less desirable than using a properly configured existing solution (like REST or JSON:API) because of the need to maintain that solution.

### Roles and Permissions considerations
This functionality would potentially be available to unathenticated clients (so as to not have the trellis service perform authentication) and the feed should be crafted in such a way that only information appropriate to anonymous users should be available.

### LINKED ISSUE #2299: Add support for Enterprise Attributes
## What is the problem that we want to solve?
As integrations are developed for Quickstart content, there are enterprise attributes associated with the Trellis ecosystem that need corresponding implementation on the Quickstart side.

This will consist of importing the attributes that exist, and allowing Quickstart content to be tagged accordingly using those attributes (notably `az_event` and `az_news`); exports of this content should provide reference to which enterprise attributes that content uses.

An example feed of the work-in-progress attributes are [available here.](https://trellis-asset.s3.us-west-2.amazonaws.com/dmo/enterprise-attributes-v2.json)

## Conditions of satisfaction

- provide a method for importing enterprise attributes
- allow quickstart content to be tagged accordingly
- provide attribute references when exporting the content

### PR DESCRIPTION
**Pull request by tadean**
## Description
This PR adds a module that contains a REST endpoint view for exporting News content to outside services.

There are two primary potential approaches this could be done with; REST and JSON:API. JSON:API is easiest, but because this data involves multiple chained references, the JSON:API output is not very easy to work with on the receiving side. To that end, this draft is attempting the REST version first, to provide output that is more directly structured.

This PR contains a view and a Serializer REST plugin that implement changing some of the references fields in the output to a more streamlined form. This approach is needed because adding true views relationships results in duplicate rows (and also doesn't provide an array structure that makes for simple output.

The JSON:API approach could be swapped to if desirable.

## Related issues
#2255 
#2299 

## How to test
- Enable `Quickstart News Export`
- Generate some News Content
- verify JSON output at `/az_quickstart/1.0/export/az_news`

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [x] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tadean** [comment]
**Todo:**

- [ ] Add `field_az_summary`

**Comment by tadean** [comment]
Added field storage for enterprise attributes to `az_core` and an `az_enterprise_attributes_import` submodule responsible for their importing.

**Comment by tadean** [comment]
- Updated the url used for the remote Trellis API endpoint for enterprise attributes to point to the latest version of the remote API. This contains human readable names for the attribute categories.
- Added custom form widget to display the enterprise attributes field as multiple form elements based on taxonomy structure:
attributes_form_element

- Added widget settings to allow selecting specific types of attributes on a per-content-type basis:
widget_settings

- Set up export API to nest attributes according to their category (e.g.):
```
"field_az_enterprise_attributes": {
  "Audiences": [
    "External Media",
    "Prospective Student"
  ],
  "Topic": [
    "Science & Research"
  ],
  "Uarizona Unit": [
    "2421"
  ]
},
```

**Comment by tadean** [comment]
- Updated to reflect new endpoint `https://api.dev.trellis.arizona.edu/ws/rest/events/v1/enterprise-attributes`
- Updated expected JSON schema to reflect enterprise attributes JSON structure changes
- Updated to allow notion of some categories of attributes having a type that influences select element (in this case, single select vs. multi-select)

**Comment by tadean** [comment]
Added a `az_enterprise_text_export` field formatter that has a trailing effect of transforming relative URLs to absolute ones, in addition to any normal filtering effects that take place due to text format.

**Comment by tadean** [comment]
Updated to add trellis team requested export schema updates.

Todo:

- [ ] Prevent edits to the enterprise attributes taxonomy vocabulary (configurable)

File: modules/custom/az_news/az_news_export/config/install/views.view.az_news_export.yml [hunk 1]
```diff
@@ -0,0 +1,1635 @@
...
+              alias: ''
+              raw_output: '1'
+            field_az_media_image:
+              alias: ''
+              raw_output: '1'
+            field_az_media_thumbnail_image:
+              alias: ''
+              raw_output: '1'
+      display_extenders: {  }
+      path: az_quickstart/1.0/export/az_news
```
Comment (joeparsons) [inline modules/custom/az_news/az_news_export/config/install/views.view.az_news_export.yml:1611]: We need to decide for sure if this is the URL pattern we want to use for this API.

**Comment by joeparsons** [comment]
My recommendation for the URL pattern for the new API endpoint would be something like this:
`az_quickstart/export/az_news/v1/az_news.json`

I think we want the "version" part of the URI to be _after_ `az_news` and I think it might be good to include the `.json` file extension.

**Comment by tadean** [comment]
> My recommendation for the URL pattern for the new API endpoint would be something like this: `az_quickstart/export/az_news/v1/az_news.json`
> 
> I think we want the "version" part of the URI to be _after_ `az_news` and I think it might be good to include the `.json` file extension.

Sounds good to me - updated the endpoint path.

**Review by joeparsons — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Pull request merged**